MAINT: remove `_lib/_c99compat.h` and use C99 rather than `npy_math.h` objects

rgommers opened the pull request:
We have required C99 for a while now, so let's remove this header. Furthermore, this PR:
- cleans up a lot of random definitions related to `isnan`/`isinf` (e.g., `sc_isinf`, `my_isinf`),
- removes the `libnpymath` dependency from `Faddeeva.cc`.

tylerjereddy reviewed (commented):
The diff looks pretty solid to me--I added one question about the "macro aliasing" in ckdtree--that might be done for concision for now I guess. Do C++ folks have strong views on "re-exporting" stdlib functions like that?

tylerjereddy commented:
To be clear, I think this is probably fine to merge as-is with green CI.

rgommers commented on scipy/spatial/ckdtree/src/ckdtree_decl.h:
```diff
@@ -5,15 +5,14 @@
 
 #define ckdtree_intp_t npy_intp
-#define ckdtree_isinf(x)   sc_isinf(x)
+#define isinf(x)   isinf(x)
```
This code is from pre-C++11 which explains why `sc_inf` was a thing. This is the minimal replacement; I could have deleted this line gone through all other source files indeed to add `from libcpp.cmath import isinf`. May be a little cleaner.

rgommers replied:
Turns out Cython's `libcpp` doesn't actually have `isinf` (or any `math`/`cmath`), so using `libc.math`.

rgommers commented:
The issue in the 32-bit Linux CI job is that that is still using GCC 5, which has C++11 support but it isn't default enabled. We need GCC >= 6.1 at least; that jumped straight to C++14 default enabled. We just bumped the minimum required GCC to 8 recently, so it's time to update that CI job anyway.

peterbell10 commented on scipy/spatial/ckdtree/src/query.cxx:
```diff
@@ -133,7 +133,7 @@ struct nodeinfo {
 
     inline void update_side_distance(const int d, const double new_side_distance, const double p) {
-        if (CKDTREE_UNLIKELY(ckdtree_isinf(p))) {
+        if (CKDTREE_UNLIKELY(isinf(p))) {
```
In C++ you should use `std::isinf`.

rgommers replied:
Okay, I'll clean this up properly.

This is what it was using before (`ckdtree_isinf` -> `sc_isinf` -> `isinf`) what I did here was just a mechanical replacement.

h-vetinari commented:
> We just bumped the minimum required GCC to 8 recently, so it's time to update that CI job anyway.

I was actually thinking of how to go about that. Should we just start with the CI on this repo? Do we expect the cibuildwheel stuff to land for 1.10?

rgommers commented:
> I was actually thinking of how to go about that. Should we just start with the CI on this repo? Do we expect the cibuildwheel stuff to land for 1.10?

Yes, `cibuildwheel` progress is quite good (see https://github.com/andyfaff/scipy/pull/28), that should be ready for the 1.10 release. So no need to bother with the `scipy-wheels` repo anymore. We also want to finish migrating all the CI jobs on this repo to Meson, which will automatically find all the cases where we're still using compilers that are too old.

rgommers commented:
Don't merge yet, the Meson on Windows job still has a weird issue.

rgommers commented:
> Don't merge yet, the Meson on Windows job still has a weird issue.

Okay, this is the second time this tricked me, so let me spell it out and then write some sort of safeguard for it somewhere:
- With `-std=c99`, `M_PI` and other `M_*` math constants are not guaranteed to be defined, _even if `_USE_MATH_DEFINES` is defined_. It is _not_ a matter of header ordering,
- On Linux and macOS these constants _are_ present if `_USE_MATH_DEFINES` is defined,
- On Windows this is _not_ the case, and https://docs.microsoft.com/en-us/cpp/c-runtime-library/math-constants#remarks is a lie,
- I verified this in the Windows GHA CI job, things were broken when using `M_SQRT1_2, and header order didn't fix it while changing `-std=c99` to `-std=gnu99` in the main `meson.build` did fix it.
- It's extra confusing that Cython lets you happily `cimport` it (see https://github.com/cython/cython/blob/0.29.x/Cython/Includes/libc/math.pxd#L16) and then it break when compiling the generated C code.

rgommers commented:
This is ready now.

h-vetinari commented on scipy/interpolate/_ppoly.pyx:
```diff
@@ -94,7 +92,7 @@ def evaluate(double_or_complex[:,:,::1] c,
             # xval was nan etc
             for jp in range(c.shape[2]):
-                out[ip, jp] = nan
+                out[ip, jp] = libc.math.NAN
```
This seems to be the only file in the diff where this isn't imported à la `from libc.math cimport NAN`

rgommers replied:
The code in this file was deliberately written like that by @pv (I'm guessing because of the `malloc`/`free` usage), so I'd much prefer not to change the style decision he made here. I chose to be consistent with the style in this file.

h-vetinari commented on scipy/spatial/src/distance_impl.h:
```diff
@@ -630,7 +628,7 @@ pdist_minkowski(const double *X, double *dm, npy_intp num_rows,
         return pdist_euclidean_double(X, dm, num_rows, num_cols);
     }
-    if (sc_isinf(p)) {
+    if (isinf(p)) {
```
Here there's no header at the top anymore, so it's not clear to me where this `isinf` is being picked up from. Perhaps add `from libc.math cimport isinf` at the top?

rgommers replied:
Good point, it should be `#include <math.h>`, will add.

h-vetinari commented on scipy/special/Faddeeva.cc:
```diff
@@ -118,34 +118,15 @@
-#endif
-#define my_isinf(x) (!my_isfinite(x) && !my_isnan(x))
+#define Inf INFINITY
+#define NaN NAN
```
These defines seem pretty pointless now. I mean, this could be a mechanical cleanup also later, but I think it'd be a good time in this PR too.

rgommers replied:
This is used quite extensively and there's upstream code, so I decided to leave it alone.

h-vetinari commented on scipy/special/_round.h:
```diff
@@ -6,15 +6,14 @@
 
-    if (sc_isnan(a) || sc_isnan(b)) {
+    if (isnan(a) || isnan(b)) {
 	return NPY_NAN;
```
Would it make sense to get rid of these numpy nans as well here?

rgommers replied:
Yes, I can do that - leads to a few more cleanups in `wright.cc` as well.

h-vetinari reviewed (commented):
LGTM overall, just some questions about consistency.

rgommers commented:
Thanks for the review @h-vetinari. I addressed the comments I thought were a good idea to add as cleanups. These changes tend to balloon and it's a bit of a puzzle, so I prefer to not add anything more to this PR unless it's necessary. There will be follow-ups later (still 27 instances of `npy_math.h` present, quite a few of which can go).

h-vetinari reviewed (approved):
LGTM!

rgommers commented:
Okay, this is happy again.

ev-br commented:
Looks good on a cursory look, I'd suggest merge and iterate if/when issues surface up.

rgommers commented:
Several thumbs up, I'll hit the green button so I can send some follow-up PRs. Thanks for the reviews everyone!
